python/ray/data/_internal/logging.py

```diff
@@ -206,3 +212,45 @@ def get_log_directory() -> Optional[str]:
 
     session_dir = global_node.get_session_dir_path()
     return os.path.join(session_dir, "logs", "ray-data")
+
+
+def get_default_formatter() -> logging.Formatter:
+    log_encoding = os.environ.get(RAY_DATA_LOG_ENCODING_ENV_VAR_NAME)
+    if log_encoding is not None and log_encoding.upper() == "JSON":
+        return DEFAULT_JSON_FORMATTER()
+
+    return logging.Formatter(DEFAULT_TEXT_FORMATTER)
+
+
+def register_dataset_logger(dataset_id: str) -> None:
+    """Add a logger for a dataset with the given ID.
+
+    Args:
+        dataset_id: The ID of the dataset.
+    """
+    global _DATASET_LOGGER_HANDLER
+    logger = logging.getLogger("ray.data")
+    if _DATASET_LOGGER_HANDLER:
+        logger.info(
+            "Concurrent dataset runs detected. The log for the current dataset will "
+            f"now be streamed to the logs of dataset {dataset_id}."
+        )
+        unregister_dataset_logger()
+
+    _DATASET_LOGGER_HANDLER = SessionFileHandler(
+        filename=f"ray-data-{dataset_id}.log",
+    )
+    _DATASET_LOGGER_HANDLER.setFormatter(get_default_formatter())
+    logger.addHandler(_DATASET_LOGGER_HANDLER)
```

Comment:
We should store both handlers in a stack and switch back to dataset1's handler, if dataset2 finishes faster. 
Otherwise, when dataset2 finishes, there is no active handler and we'll loose the future logs for dataset1

Reply:
and can you also add some unit tests that cover different scenarios? 
E.g, 
* sequential
* concurent, dataset2 finishes first
* concurrent, dataset1 finishes first.